api/paidAction/index.js

```
@@ -41,243 +41,246 @@ export default async function performPaidAction (actionType, args, incomingConte
       }
-
-      // payment methods that anonymous users can use
-      if (paymentMethod === PAID_ACTION_PAYMENT_METHODS.P2P) {
-        try {
-          return await performP2PAction(actionType, args, context)
-        } catch (e) {
-          if (e instanceof NonInvoiceablePeerError) {
-            console.log('peer cannot be invoiced, skipping')
-            continue
-          }
-          console.error(`${paymentMethod} action failed`, e)
-          throw e
-        }
-      } else if (paymentMethod === PAID_ACTION_PAYMENT_METHODS.PESSIMISTIC) {
-        return await beginPessimisticAction(actionType, args, context)
+      const forcedPaymentMethods = []
+      if (context.supportedPaymentMethods.includes(PAID_ACTION_PAYMENT_METHODS.FEE_CREDIT)) {
+        forcedPaymentMethods.push(PAID_ACTION_PAYMENT_METHODS.FEE_CREDIT)
+      }
+      if (forcedPaymentMethods.length === 0) {
+        throw new Error('action does not support internal payments')
       }
+      context.supportedPaymentMethods = forcedPaymentMethods
+    } else if (prioritizeInternal) {
+      // prefer internal payment methods
+      const priority = {
+        [PAID_ACTION_PAYMENT_METHODS.FEE_CREDIT]: -2
+        // add other internal methods here
+      }
+      context.supportedPaymentMethods = context.supportedPaymentMethods.sort((a, b) => {
+        return priority[a] - priority[b]
+      })
+    }
+
+    const { supportedPaymentMethods } = context
 
-      // additionalpayment methods that logged in users can use
-      if (me) {
-        if (paymentMethod === PAID_ACTION_PAYMENT_METHODS.FEE_CREDIT) {
-          try {
+    for (const paymentMethod of supportedPaymentMethods) {
+      try {
+        console.log(`trying payment method ${paymentMethod}`)
+        switch (paymentMethod) {
+          case PAID_ACTION_PAYMENT_METHODS.FEE_CREDIT: {
+            if (!me || (me.msats ?? 0n) < cost) break // if anon or low balance skip
```

The original code that looks for `violates check` was there for a reason. The way you've rewritten this suppresses errors that aren't related to payments.

The switch statement is nice, but be careful causing regressions by changing code that you don't understand - and more importantly, code that isn't directly related to your changes. 

Stuff like this is what causes me to do full rewrites.